qmcpy/discrete_distribution/lattice/lattice.py

```diff
@@ -132,93 +144,137 @@ def __init__(self, dimension=1, randomize=True, order='natural', seed=None,
             If `generating_vector` is a string (path), `d_max` and `m_max` can be taken from the file name if None.
         """
-        self.parameters = ['dvec','randomize','order']
-        self.randomize = randomize
-        self.order = order.lower()
-        if self.order == 'natural':
-            self.gen = self._gail_natural
-        elif self.order == 'linear':
-            self.gen = self._gail_linear
-        elif self.order == 'mps':
-            self.gen = self._mps
-        else:
-            raise Exception("Lattice requires natural, linear, or mps ordering.")
+        self.parameters = ['dvec','randomize','order','gen_vec']
+        self.mimics = 'StdUniform'
+        self.low_discrepancy = True
+        self.replications_gv = 1
+        # ordering
+        self.order = order.upper()
+        if self.order=="MPS": raise ParameterError("lattice in MPS order is no longer supported")
+        assert self.order in ['LINEAR','NATURAL','GRAY']
+        # generating vector
         if isinstance(generating_vector,ndarray):
             self.gen_vec_og = generating_vector
             if d_max is None or m_max is None:
                 raise ParameterError("d_max and m_max must be supplied when generating_vector is a ndarray")
             self.d_max = d_max
             self.m_max = m_max
+            if self.gen_vec_og.ndim==1:
+                self.gen_vec_og = self.gen_vec_og[None,:]
+            else:
+                assert self.gen_vec_og.ndim==2, "ndarray generating vector must be 1 or two dimensional"
+                self.replications_gv = len(self.gen_vec_og)
         elif isinstance(generating_vector,int):
             self.m_max = min(26,max(2,generating_vector))
+            assert isinstance(dimension,int), "random generating vector requires int dimension"
             self.d_max = dimension
         elif isinstance(generating_vector,str):
             parts = generating_vector.split('.')
             root = dirname(abspath(__file__))+'/generating_vectors/'
             repos = DataSource()
             if isfile(root+generating_vector):
-                self.gen_vec_og = load(root+generating_vector).astype(uint64)
+                self.gen_vec_og = load(root+generating_vector).astype(uint64)[None,:]
                 self.d_max = int(parts[-3])
                 self.m_max = int(parts[-2])
             elif isfile(generating_vector):
-                self.gen_vec_og = load(generating_vector).astype(uint64)
+                self.gen_vec_og = load(generating_vector).astype(uint64)[None,:]
                 self.d_max = int(parts[-3])
                 self.m_max = int(parts[-2])
-            elif "LDData"==generating_vector[:6] and repos.exists("https://raw.githubusercontent.com/QMCSoftware/"+generating_vector):
-                datafile = repos.open("https://raw.githubusercontent.com/QMCSoftware/"+generating_vector)
+            elif "LDData"==generating_vector[:6]:
```

Comment:
Is there any way we can refactor this out into the package? Relying on a download could fail if GitHub is down.

Reply:
The default generating vector for lattices and generating matrices for digital nets are distributed with the package and do not require an internet connection to access. I added cases to `lattice.py` and `digital_net_b2.py` that first check for a local path to the file for those who have cloned `QMCSoftware/LDData/` repo to use offline. The case you highlighted will download the file for users if it is not available locally given that they have have internet access and GitHub is up and running.